### PR TITLE
store-gateway: fix TSDB index v1 symbols lookup and caching

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
#### Context

* The symbol reference in prometheus' TSDB v1 index is an offset into the index file. In V2 it is the sequence number of the symbol.

#### What this PR does

While working on https://github.com/grafana/mimir/pull/5073 I noticed that the symbols offsets are slightly misaligned for V1 symbols. 

This PR fixes three bugs

* label names and values symbols caches were misaligned
	* __label names cache__ was using the index header offset to store the cache items, but using the TSDB index to look them up (difference of 9)
		* This results in mostly cache misses; but it also interacts with the other bug of returning symbols references relative to the start of the symbols table, not the whole index, so there can be incorrect values returned as well
	* __label values cache__ was using the TSDB offset to lookup and store cache items, but collision detection was using the index header offset; 
		* this results is 100% cache misses for label values
* label names cache was also populated incorrectly
	* the offsets there were offsets relative to the start of the symbols table, not relative to the index header or the TSDB index (difference of 5)

Here is some ascii art to better illustrate the implications of these bugs

```
offset in TSDB:          320 // this was the cache key for looking up label names 
offset in index header:  329 // the index header has 9 more bytes in its header than the TSDB index
offset in symbols table: 315 // excludes the TSDB index header of 5 bytes (magic+version); this was the cache key for storing label names


|__<8B>__|__<5B>_|________...___4fooo3bar
^        ^       ^              ^    ^
|        |       |              |    |
start of index header           |    |
         |       |              |    |
         start of TSDB index    |    |
                 |              |    |
                 start of symbols table
                                |    |
                                symbol with TSDB offset 315, length 4 and string fooo; this is the cache key for label names; this is looked up in the cache as symbol ref 315
                                     |
                                     symbol with TSDB offset 320, length 3 and string bar; this is cached as symbol ref 315
                                      
```

Testing - I am not aware of a way to generate a V1 index (maybe checking out prometheus at the time and running some code, sounds complicated?), so the changes wrt caching are untested. Only the changes in the Symbols struct covered by tests.

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
I took a quick look. These changes are safe with regards index v2, which is what we use. I doubt we have any OSS user using v1, reason why I also mentioned to you I would be keen to deprecate it. Until then, happy you fixed it!

File: pkg/storegateway/indexheader/index/symbols_test.go [hunk 1]
```diff
@@ -25,7 +27,91 @@ func TestMain(m *testing.M) {
...
+	t.Run("Lookup", func(t *testing.T) {
+		for ref, symbol := range refToSymbol {
+			s, err := s.Lookup(ref)
+			require.NoError(t, err)
+			require.Equal(t, symbol, s)
+		}
+
+		// Look beyond the last symbol
+		_, err = s.Lookup(uint32(symbolsTableSize + len(filePrefix) + 1))
+		require.Error(t, err)
```
Comment (charleskorn) [inline pkg/storegateway/indexheader/index/symbols_test.go:87]: Can we be more specific about which error we expect here? (same comment applies to other `require.Error` calls)

**Review by charleskorn — APPROVED** [review]
LGTM.

> Testing - I am not aware of a way to generate a V1 index (maybe checking out prometheus at the time and running some code, sounds complicated?), so the changes wrt caching are untested. Only the changes in the Symbols struct covered by tests.

There's a v1 index in `pkg/storegateway/indexheader/testdata/index_format_v1` - can we use that?

Also, there's `TestReadersComparedToIndexHeader` in `pkg/storegateway/indexheader/header_test.go` that checks the symbols in an index header for a v1 index - is this test correct? Does it need to be expanded?

**Comment by dimitarvdimitrov** [comment]
> There's a v1 index in `pkg/storegateway/indexheader/testdata/index_format_v1` - can we use that?

I saw that but the label names there aren't susceptible to this bug.



> Also, there's `TestReadersComparedToIndexHeader` in `pkg/storegateway/indexheader/header_test.go` that checks the symbols in an index header for a v1 index - is this test correct? Does it need to be expanded?

The bug here is in caching and usually results in cache misses. Correctness is only affected in some slim  cases, in the rest only performance is affected. That's why `TestReadersComparedToIndexHeader` passes. So this change can be seen in benchmarks for Symbols reading in V1; I couldn't find any such benchmarks.

File: pkg/storegateway/indexheader/index/symbols_test.go (hunk 1 above)
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/index/symbols_test.go]: it required more massaging than I expected, but take a look at https://github.com/grafana/mimir/pull/5103/commits/508efb9ba36d333f300821cf66059cc69b83b397

**Comment by dimitarvdimitrov** [comment]
I will merge this because I want to have it in r241. There's nothing critical in the remaining comments, and I'm happy to open follow-ups if we come up with good ways to test this

**Pull request merged**